File: removeNthnodefromlist.py
# leetcode: accepted
# time complexity: o(n)
# doubts (yes):  please check the comments
# explaination: we are taking two pointers, p1 and p2. We shift the p1 pointers , n places from the begining and then
# we move the p1 till the end if the linkedlist and move p2 along with it.

# when p1 shows null, it means p2 is n positions from end of linkedlist. (the difference between p1 and p2 is n) so when the p1 shows null, it means its n'th position from end of linkedlist

# aftet that we just link the p2 to p2.next.next and return the list

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        p1 = head
        p2 = head

        for i in range(n):
            p1 = p1.next
        if not p1:  # what does this line do
            return head.next
        while (p1.next != None):  # it also gives wrong o/p when i put p1!=None
            p2 = p2.next  # the output is wrong when i traverse p1 first and then traverse p2
            p1 = p1.next

        # print(p2.val)
        p2.next = p2.next.next
        return head

File: Remove_Nth_Node_From_List.java
// Time Complexity :O( n)
// Space Complexity : O(1) 
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No



/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
    ListNode dummy = new ListNode(-1);
    dummy.next = head;
    ListNode slow=dummy;
    ListNode fast = dummy;
        
    for(int i=0;i<n;i++){
        fast = fast.next;
    }
    // System.out.println(slow.val +"fast"+ fast.val);
    while(fast.next !=null){
        slow = slow.next;
        fast=fast.next;}
   
    slow.next = slow.next.next;
//    ListNode temp= slow.next;
//     slow.next = temp.next;    
//     temp.next = null;    
    
    return dummy.next;
}
}
